adr/2023-01-31-infrastructure-architecture.md

```diff
@@ -0,0 +1,91 @@
+1. CloudFront will further improve the availability and performance of the API by caching redirect responses and serving responses geographically close to the user.
+1. The use of Lambda Function URLs entirely removes the need to manage an API Gateway instance.
+1. DynamoDB will provide a highly available and scalable data store for the API.
+1. There is not currently a need for a persistent service, so the value of ECS in that regard is negated.
+1. By avoiding the use of ECS, we also do not need to include a load balancer in our architecture.  This further reduces the request bottlenecks and improves the scalability of the API.
+
+![URL Shortener API infrastructure diagram](attachments/2023-01-31-infrastructure-architecture.png)
+__Figure 1:__ URL Shortener API infrastructure diagram ([source](attachments/2023-01-31-infrastructure-architecture.drawio))
+
+## Consequences
+
+### DynamoDB data model
+
+We will need to be careful in how we design our DynamoDB item data model.  It is imperative that we do not create a situation where data cannot be directly retrieved and must instead by scanned for.  The reason being that DynamoDB scan operations are expensive and inefficient.
+
+### Persistent service limitation
+
+If we discover a future requirement that needs a persistent service, we will need to add additional infrastructure to support it.  As persistent services are generally only needed to run scheduled or long-running jobs, this should not impact us:
+
+- Scheduled jobs can be handled with AWS CloudWatch Events and AWS Lambda.
+- Long-running jobs would have to exceed 15 minutes before they became problematic for the Lambda.  If they did, we could then look at using a Step Function or introducting a dedicated ECS task that is triggered via a CloudWatch Event.
+
+### Lambda cold-start time
+
+A limitation of Lambda is that on a function's first invocation, it will take longer to execute and return a response than on subsequent invocations.  If we discover that the API's Lambda function cold-start time is impacting user experience, we can mitigate with a few different startup strategies:
+
+- Use a health check to keep Lambda function's warm.
+- Use Lambda provisioned concurrency to always have a configurable nubmer of Lambda functions warm and ready to serve requests.
+
+### Preventing malicious network egress
+
+There is a requirement for the API to check shortened URLs are still valid.  This will be accomplished by making a HEAD request to the full URL and considering it invalid if the response is in the 4xx HTTP status code range.  
+
+This will require the API to perform DNS looksup and make network requests to external URLs.  To mitigate the risk of malicious network egress, Route53 DNS safelisting will be used to only allow the API to perform DNS lookups for domains that are part of the service's allow list (e.g. `*.canada.ca` and `*.gc.ca`).
```

Comment:
Is there value/possible to only allow the dsnlookup and no other protocols?

Reply:
We'll be able to lock it down by port and protocol using security groups.  This will let us say things like:

> Only allow the lambda to perform TCP egress on port 443 to specific IP addresses

In our case, the "specific IP addresses" will end up being the entire internet though using `0.0.0.0/0` 😅